### PR TITLE
feat(engine): structured output — JSON constrained decoding (Track C)

### PR DESCRIPTION
**Pull request by magicnight**
## Summary
The v0.6 three-pack's differentiator: `response_format` support with a hand-written byte-level JSON grammar compiled to a logit mask — **zero third-party dependencies** (mlx-lm ships none built in; oMLX depends on external xgrammar).

- **C1** `{"type":"json_object"}` — byte-level pushdown automaton constrains output to well-formed JSON (RFC 8259 numbers/strings/escapes with surrogate-pair enforcement, depth cap).
- **C2** `{"type":"json_schema"}` subset — top-level object, string/number/integer/boolean/string-enum properties, `required`, any-order keys. Everything else is a **400** ("unsupported schema feature/keyword"): strict whitelist, nothing silently downgraded (incl. `pattern`/`minimum`/`format`); keys/enum values requiring JSON escaping or non-ASCII are rejected at compile time so the automaton can never enter an unmatchable state.
- **Masking**: upstream `LogitProcessor` on a hand-built `TokenIterator` (streaming path byte-identical to normal); penalties first, constraint last; greedy argmax-first fast path provably equivalent to full-mask argmax; on-GPU one-hot. **All-illegal states log a diagnostic and force EOS** — the well-formed-JSON promise is never silently broken.
- Per-model token→bytes table built once and cached; batch/speculative paths excluded per-request; Track B prompt cache composes safely.

## Review & verification
- Adversarial review (opus): REQUEST CHANGES → the HIGH (all-illegal fallback returned unmasked logits = silent garbage) fixed with forced-EOS + diagnostic and locked by Metal mask tests; M1/M2 decoder guards (keyword whitelist, unmatchable key/enum rejection) + M3 (processor decision logic extracted MLX-free into CI) + surrogate-pair enforcement all landed.
- Verification: 64 MLX-free tests; gated Metal mask tests 3/3; **real-model E2E on Qwen3-0.6B/4B: C1 100% JSONSerialization-parseable, C2 100% schema-conformant**; final full double-run 378/48 suites green (xcodebuild + bare).
- Throughput overhead (informational): ~6% on 4B, 9-21% variance on 0.6B (fixed mask cost dominates on small/fast models) — perf follow-up registered.

## Track status
This completes the v0.6 three-pack: **A continuous batching ✅ · B prefix cache ✅ · C structured output (this PR)**. Remaining for release per the strict-three-pack decision: agent real-world benchmark.

---

> [!NOTE]
> **Medium Risk**
> Large new inference path touching every decode step and HTTP validation; correctness is heavily tested but wedged automaton / forced-EOS behavior affects client-visible output and finish reasons.
> 
> **Overview**
> Adds **Track C structured output**: OpenAI `response_format` on chat completions, wired through validation, generation, and routing.
> 
> **API & request path:** `HummingbirdServer` accepts `response_format`; `ResponseFormatDecoder` compiles `json_object` / a strict `json_schema` subset into `ResponseFormat` on `GenerateRequest`, returning **400** for unsupported keywords, nested types, or schema keys/enums the byte matcher cannot spell. Unsupported features are not silently ignored.
> 
> **Constrained decoding:** New byte-level automata (`JSONGrammarState` for any JSON, `SchemaConstraintState` for flat object schemas) sit behind `ConstraintState`. `JSONConstraintProcessor` masks illegal vocabulary tokens each step (penalties first, constraint last), with greedy top-K fast path and full-vocab masking when sampling. Per-model `TokenVocabularyTable` is built once and cached. When no legal token exists, the processor **logs and forces EOS** instead of emitting unmasked logits.
> 
> **Engine integration:** `MLXSwiftEngine` uses a hand-built `TokenIterator` with the constraint processor on the single-stream LLM path when `responseFormat` is set. **Batching and speculative decoding are disabled** for constrained requests (`BatchRoutingPolicy`, draft branch skipped). `LogManager.logSync` supports synchronous logging from the logit hot path.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0d5710ec77265ee82278d3814c118795a8562086. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: MacMLXCore/Sources/MacMLXCore/Engine/MLXSwiftEngine.swift [hunk 1]
```diff
@@ -972,12 +1062,31 @@ public actor MLXSwiftEngine: InferenceEngine {
...
+                        vocabCache: vocabCache
+                    )
+                } else {
+                    stream = try MLXLMCommon.generateTokens(
+                        input: inputBox.value,
+                        cache: cache,
+                        parameters: generateParams,
+                        context: context
+                    )
+                }
```
[the remark this answers is not in the document]
Comment (magicnight) [reply MacMLXCore/Sources/MacMLXCore/Engine/MLXSwiftEngine.swift:1089]: Verified: confirmed. The combination was admitted (only logprobs+response_format and logprobs+draft were 400'd) and for models whose toolCallFormat is .json the constrained document's leading '{' buffers the whole response in ToolCallProcessor — and can be swallowed as a tool call. Fixed in #83: tools+response_format now returns 400 at the HTTP layer (consistent with the other unsupported combos), and makeToolProcessor additionally gates on responseFormat as defense-in-depth for non-HTTP paths.

File: MacMLXCore/Sources/MacMLXCore/Engine/MLXSwiftEngine.swift [hunk 2]
```diff
@@ -823,11 +829,77 @@ public actor MLXSwiftEngine: InferenceEngine {
                 modelID: loadedModelSnapshot.id,
                 hasTools: request.tools?.isEmpty == false,
                 numDraftTokens: request.numDraftTokens,
+                responseFormat: request.responseFormat,
```
[the remark this answers is not in the document]
Comment (magicnight) [reply MacMLXCore/Sources/MacMLXCore/Engine/MLXSwiftEngine.swift:832]: Verified: confirmed at High. The VLM branch logged every other ignored feature but applied response_format silently-unconstrained — a direct violation of this project's no-silent-degradation rule. Fixed in #83: images + response_format now returns 400 before generation.
